### PR TITLE
[PyTorch] Add no-load-balance THD all-gather CP

### PR DESCRIPTION
**Pull request by sudhakarsingh27**
## How to use

Select the token partition strategy explicitly and use the same value for both
context-parallel attention configuration and input partitioning:

```python
from transformer_engine.pytorch import CPAttentionLoadBalancingStrategy

strategy = CPAttentionLoadBalancingStrategy.NO_LOAD_BALANCE

layer.set_context_parallel_group(
    cp_group,
    cp_global_ranks,
    cp_stream,
    cp_comm_type="all_gather",
    load_balancing_strategy=strategy,
)

input_ids, labels, position_ids = get_batch_on_this_cp_rank(
    cu_seqlens_padded,
    input_ids,
    labels,
    position_ids,
    cp_group=cp_group,
    load_balancing_strategy=strategy,
)
```

`CPAttentionLoadBalancingStrategy.DUAL_CHUNK_SWAP` remains the default.

## What changed

- Add the public `CPAttentionLoadBalancingStrategy` enum with
  `DUAL_CHUNK_SWAP` and experimental `NO_LOAD_BALANCE` strategies.
- Propagate the selected strategy through TransformerLayer,
  MultiheadAttention, DotProductAttention, and the context-parallel attention
  backends.
- Assign one contiguous physical-token chunk to each CP rank for
  `NO_LOAD_BALANCE`, while preserving logical document boundaries through THD
  metadata.
- Capture the selected strategy for backward so forward and backward use the
  same token layout.
- Reuse the existing native THD partition-index implementation for CUDA batch
  slicing while retaining the CPU dataloader fallback.
- Add focused utility, propagation, FusedAttention, and gated unpadded
  FlashAttention 3 coverage.

## Why

The default per-document DualChunkSwap partition divides every sequence into
`2 * cp_size` chunks and performs two attention steps per rank. Partitioning
the complete physical buffer into `cp_size` contiguous chunks allows one
attention step per rank and supports packed documents whose individual lengths
are not divisible by `2 * cp_size`.

This strategy intentionally trades causal load balance for fewer attention
calls.

## Scope and constraints

The experimental strategy currently requires THD self-attention,
`cp_comm_type="all_gather"`, causal attention without a sliding window
(`window_size=(-1, 0)`), equal local Q/K/V physical lengths, and either
FusedAttention or FlashAttention 3. FlashAttention 3 requires
`pad_between_seqs=False`. FP8 and CUDA graph capture are not supported.

Input partitioning and attention configuration must use the same strategy.

## Validation

- 21 focused context-parallel utility tests passed.
- CP2 BF16 THD all-gather FusedAttention forward/backward passed for both
  `DUAL_CHUNK_SWAP` and `NO_LOAD_BALANCE`.
- TransformerLayer strategy propagation and the default
  `DUAL_CHUNK_SWAP` behavior passed focused smoke checks.
- Autograd forward/backward arity validation and `git diff --check` passed.
- Pylint passed with a 10.00/10 score on the changed production files.

### CODE REVIEW

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 1]
```diff
@@ -56,6 +56,14 @@
 
 # Float8CurrentScaling: fused_attn_bwd takes O in FP8 by default, this flag allows it in F16
 _dpa_fp8_cs_o_in_f16 = os.getenv("NVTE_DPA_FP8CS_O_in_F16", "1") == "1"
+_NO_LOAD_BALANCE_ENV = "NVTE_EXPERIMENTAL_CP_AG_THD_NO_LOAD_BALANCE"
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: Jax CP defines "DualChunkSwap", "Stripped", and "NoLoadBalance" as some Strategy (an enum). Maybe we can do the same on the PyTorch side? We can still label it experimental that way (in the docs), but without the environment variable.

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 2]
```diff
@@ -3073,6 +3201,30 @@ def forward(
...
+            assert not fp8, "No-load-balance THD partitioning does not support FP8 yet."
+            assert (
+                not is_graph_capturing()
+            ), "No-load-balance THD partitioning does not support CUDA graph capture yet."
+            assert (
+                q.shape[0] == k.shape[0] == v.shape[0]
+            ), "No-load-balance THD partitioning requires equal local Q/K/V physical lengths."
+            assert cu_seqlens_q is cu_seqlens_kv and (
+                cu_seqlens_q_padded is cu_seqlens_kv_padded
+            ), "No-load-balance THD self-attention requires shared Q/KV sequence metadata tensors."
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: shouldn't these checks be in `attn_forward_func_with_cp` - there's already one in there?

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 3]
```diff
@@ -3056,6 +3184,7 @@ def forward(
         fp8_meta,
         quantizers,
         fp8_output,
+        no_load_balance,
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: These "no_load_balance" flags in code and tests can all become an Enum, or a flag for `load_balancing_strategy` I think.

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 4]
```diff
@@ -3231,12 +3363,8 @@ def forward(
...
             k_ag = k_ag.view(-1, *k.shape[1:])
             v_ag = v_ag.view(-1, *v.shape[1:])
-        # cp_stream is used for step 1 of the per-step loop and must wait until
-        # k_ag/v_ag preparation finishes on the current stream — otherwise step 1
-        # races against AG/reorder writes. Manifests at high cp_size where reorder
-        # is large enough to outlast cp_stream's launch (e.g. bucket128k @ cp=8).
-        cp_stream.wait_stream(torch.cuda.current_stream())
-
+            # Preserve overlap by letting cp_stream proceed before output initialization.
+            cp_stream.wait_stream(torch.cuda.current_stream())
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py:3375]: Why shifting `wait_stream` to only the `else` branch? Any particular reason for this?

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 5]
```diff
@@ -4959,10 +5117,18 @@ def attn_forward_func_with_cp(
...
+    implementation applies DualChunkSwap to each sequence independently. It requires
+    every sequence length to be, or be padded to be, divisible by (cp_size * 2), and
+    tokens must be re-ordered before entering this function.
+
+    Experimental environment flag
+    ``NVTE_EXPERIMENTAL_CP_AG_THD_NO_LOAD_BALANCE=1`` instead assigns one
+    contiguous physical-buffer chunk to each rank and uses one attention step per rank.
+    Logical sequences remain isolated by ``cu_seqlens``. This mode requires THD,
+    all-gather, full causal self-attention, and FusedAttention, or FlashAttention 3
+    without padding. Input producers must use the same flag when partitioning inputs
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: I guess "without padding" here means `pad_between_seqs=False`?

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 6]
```diff
@@ -5040,6 +5206,31 @@ def attn_forward_func_with_cp(
...
+        assert (
+            use_fused_attention or use_flash_attn_3
+        ), "No-load-balance THD partitioning requires FusedAttention or FlashAttention 3."
+        assert not (
+            use_flash_attn_3 and pad_between_seqs
+        ), "No-load-balance THD partitioning with FlashAttention 3 does not support padding yet."
+        assert "causal" in attn_mask_type and window_size == (
+            -1,
+            0,
+        ), "No-load-balance THD partitioning currently supports full causal attention only."
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: I think you mentioned somewhere else about "full causal attention". I understand you mean "no sliding window", but I feel generally when people say causal mask or causal attention, they do mean "full causal"? (nit)

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 7]
```diff
@@ -5277,24 +5471,57 @@ def get_batch_on_this_cp_rank(
...
+                    )
+
+            else:
+                total_slices = 2 * cp_size
+                slice_sizes = (cu_seqlens_padded[1:] - cu_seqlens_padded[:-1]) // total_slices
+
+                def build_rank_indices(device):
+                    # Preserve the CPU-capable per-document dataloader path.
+                    rank_slices = []
+                    for slice_size, seq_start in zip(slice_sizes, cu_seqlens_padded[:-1]):
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py:5552]: Is there any clever way to do this without this loop? Can you ask your agent? :)

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 1 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: Done. The environment flag is replaced by the public `CPLoadBalancingStrategy` enum. `NO_LOAD_BALANCE` is documented as experimental, and `DUAL_CHUNK_SWAP` remains the default.

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 3 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: Done. The booleans are replaced with `CPLoadBalancingStrategy`, propagated through the attention stack, and captured for backward.

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 4 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py:3375]: The THD wait follows K/V restoration and metadata preparation; the non-THD wait remains after K/V reordering. Comments now clarify both dependencies.

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 5 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: Yes—this specifically means `pad_between_seqs=False`. The documentation and validation now state that explicitly.

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 6 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: Updated to “causal attention without a sliding window,” specifically `window_size=(-1, 0)`.

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 7 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py:5552]: Done. CUDA partitioning now uses `tex.thd_get_partitioned_indices`; the Python loop remains only as the CPU fallback.

**Review by cyanguwa — DISMISSED** [review]

File: tests/pytorch/attention/test_cp_utils.py [hunk 8]
```diff
@@ -9,14 +9,16 @@
 import unittest
 from transformer_engine.pytorch import CPLoadBalancingStrategy
 from transformer_engine.pytorch.attention.dot_product_attention.context_parallel import (
-    get_no_load_balance_thd_causal_metadata,
+    _zero_thd_padding,
```
Comment (cyanguwa) [inline tests/pytorch/attention/test_cp_utils.py:12]: How come we only added this now? Without this, how did the CI pass before?

File: tests/pytorch/attention/test_attention_with_cp.py [hunk 9]
```diff
@@ -716,12 +709,20 @@ def test_cp_with_fused_attention(
 )
 def test_cp_with_fused_attention_no_load_balance(cp_pool):
     """Check experimental single-chunk forward/backward."""
-    # cp_2_0 reaches the generic test's known deterministic THD backward OOM threshold on sm90.
-    if _deterministic and get_device_compute_capability() == (9, 0):
-        pytest.skip(
-            "Deterministic FusedAttention backward with THD format OOMs on sm90"
-            " for large bHSS configs (known cuDNN issue)."
-        )
+    config = copy.deepcopy(model_configs_fused_attn["cp_2_0"])
```
Comment (cyanguwa) [inline tests/pytorch/attention/test_attention_with_cp.py:712]: I just noticed that we are only testing cp_2_0 config. It is pretty representative of quite a few features, but is this intended to be just one test? (Doesn't have to add more, but just curious.)

File: transformer_engine/pytorch/attention/dot_product_attention/utils.py [hunk 10]
```diff
@@ -1565,6 +1559,26 @@ def _is_fa3_supported(num_heads, num_gqa_groups, head_dim_qk, head_dim_v, qkv_dt
...
+        # grow super-linearly (B=4 took ~4x the B=2 amount, not 2x) — revisit if a
+        # config uses B>2.
+        SM90_DET_FUSED_THD_BWD_MAX_BHSS = 1 << 30
+        if (
+            use_fused_attention
+            and is_training
+            and qkv_format == "thd"
+            and device_compute_capability == (9, 0)
+            and batch_size * num_heads * max_seqlen_q * max_seqlen_kv
+            >= SM90_DET_FUSED_THD_BWD_MAX_BHSS
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/utils.py:1575]: Should we add a condition here so it's only for "FusedAttnBackend.F16_arbitrary_seqlen" backend?

**Review by cyanguwa — DISMISSED** [review]
Two minor comments - please run the local CI at least and I'm happy to approve again. Thanks for the PR!

File: tests/pytorch/attention/test_attention_with_cp.py (hunk 9 above)
Comment (sudhakarsingh27) [reply tests/pytorch/attention/test_attention_with_cp.py:712]: Yes, one representative end-to-end case is intentional. Parameterizing `load_balancing_strategy` across the existing 12,288-case essential matrix would double collection, with most added cases skipped because no-load-balance currently supports only THD/all-gather/causal/no-SWA/non-FP8. Partitioning and metadata edge cases are covered separately in `test_cp_utils.py`.

File: tests/pytorch/attention/test_cp_utils.py (hunk 8 above)
Comment (sudhakarsingh27) [reply tests/pytorch/attention/test_cp_utils.py:12]: This isn't part of the no-load-balance change. The helper, import, and tests landed together in #2994 and entered this branch when we merged current `main`; before that, neither the helper nor those tests was present.

File: transformer_engine/pytorch/attention/dot_product_attention/utils.py (hunk 10 above)
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/attention/dot_product_attention/utils.py:1575]: Yes. I restricted the guard to FusedAttnBackend.F16_arbitrary_seqlen; the workspace issue is specific to deterministic F16/BF16 THD backward, so FP8 is no longer disabled. I verified the SM90 boundary and control cases and confirmed SM100 remains unaffected.

**Review by cyanguwa — APPROVED** [review]
